pkg/klusterlet/clusterclaim/clusterclaim_controller.go

```diff
@@ -4,98 +4,303 @@ import (
+		} else if len(subs) > 2 {
+			name = strings.ReplaceAll(name, "/", ".")
+		}
+		name = strings.ReplaceAll(name, "_", "-")
+
+		// ignore the label if the transformed name is still not a valid resource name
+		if errs := validation.IsDNS1123Subdomain(name); len(errs) > 0 {
+			klog.V(4).Infof("skip syncing label %q of ManagedClusterInfo to ClusterCliam because it's an invalid resource name", label)
 			continue
 		}
-		err := r.ClusterClient.ClusterV1alpha1().ClusterClaims().Delete(ctx, claim.Name, metav1.DeleteOptions{})
-		if err != nil {
-			errs = append(errs, err)
+
+		// ignore the label if its value is empty. (the value of ClusterCliam can not be empty)
+		if len(value) == 0 {
+			klog.V(4).Infof("skip syncing label %q of ManagedClusterInfo to ClusterClaim because its value is empty.", label)
+			continue
 		}
+
+		claim := newClusterClaim(name, value)
+		if claim.Labels != nil {
+			claim.Labels[labelCustomizedOnly] = ""
+		}
+		claims = append(claims, claim)
 	}
+	return claims, nil
+}
 
-	return utils.NewMultiLineAggregate(errs)
+func filterOutStableClaims(claims []*clusterv1alpha1.ClusterClaim) (filtered []*clusterv1alpha1.ClusterClaim) {
+	for _, c := range claims {
+		if !stableClaims.Has(c.Name) {
+			filtered = append(filtered, c)
+		}
+	}
+	return filtered
+}
+
+type updateCheck func(oldClaim, newClaim *clusterv1alpha1.ClusterClaim) bool
+
+// There are 3 cases that we don't want to update ClusterClaim:
+//  1. newClaim.Name is in clusterClaimCreateOnlyList
+//  2. oldClaim.Name is ClaimOCMProduct and oldClaim.Spec.Value is not empty
+//     and newClaim.Spec.Value is ProductOther
+//  3. oldClaim.Name is ClaimOCMPlatform and oldClaim.Spec.Value is not empty
+//     and newClaim.Spec.Value is PlatformOther
+//
+// In a word, for product and platform, we don't want update them from a specific value to "Other".
+// This is because in the process of upgrading, platform and product could be detected as "Other"
+// then reflect on the console.
+var updateChecks = []updateCheck{
+	func(oldClaim, newClaim *clusterv1alpha1.ClusterClaim) bool {
+		return !utils.ContainsString(clusterClaimCreateOnlyList, newClaim.Name) &&
+			oldClaim.Spec.Value != ""
+	},
+	func(oldClaim, newClaim *clusterv1alpha1.ClusterClaim) bool {
+		// only check ClaimOCMProduct
+		if oldClaim.Name != ClaimOCMProduct {
+			return true
+		}
+		// don't allow update from a specific value to "Other"
+		if newClaim.Spec.Value == ProductOther {
+			return false
+		}
+		// do not allow update unless from other
+		if oldClaim.Spec.Value != ProductOther {
+			return false
+		}
+		return true
+	},
+	func(oldClaim, newClaim *clusterv1alpha1.ClusterClaim) bool {
+		// only check ClaimOCMProduct
+		if oldClaim.Name != ClaimOCMPlatform {
+			return true
+		}
+		// don't allow update from a specific value to "Other"
+		if newClaim.Spec.Value == PlatformOther {
+			return false
+		}
+		// do not allow update unless from other
+		if oldClaim.Spec.Value != PlatformOther {
```

Comment:
@qiujian16 @zhiweiyin318 I think we should delete this because, in the rosa-host case, the configmap of rosa-logo has the chance of being created after the pod runs for a while. 

It will determine product as OCP first, but we want it turns to ROSA later.